hathitrust/ht_full_text_search
# Setting up Solr server using authentication and updating the python code to access solr using user and password

This PR updates the docker-compose and the Python code to start up and access to a Solr server using authentication. 

The pull request includes:

Updating documentation
Fixing all the tests.
We are updating the Python code to access Solr with authentication.
Updating docker file and docker-compose to start up Solr with authentication 
In the docker-compose has been included a service to index some data in Solr (Useful for testing)

**How do we test this PR?**

```
git checkout readme_improvements
./init.sh 
docker compose up -d
```

**Check if the API is running Access to the URL:** http://localhost:8000/docs

**Run the tests**
`docker compose exec full_text_searcher python -m pytest`

You will see the following output if all the tests ran without issue

<img width="1508" alt="image" src="https://github.com/user-attachments/assets/773fb982-7682-4d80-ab7d-34c5a5b67bb7">


**[Optional] In case you want to run any of the implemented features**

**Run the use case 1: Query the Solr index given a phrase**

```
docker compose exec full_text_searcher python ht_full_text_search/ht_full_text_searcher.py --env dev --query_string "persons and corporations" --query_config ocronly
```

Output: In your terminal, you should see the list of documents obtained with the input query.

<img width="1505" alt="image" src="https://github.com/user-attachments/assets/7d52a662-f23d-4c11-ab6d-9305138661b7">

**Run the Use case 2: Query Solr given a list of phases (included in list_query_file.txt) and generate an CSV file with the query results.**

```
docker compose exec full_text_searcher python scripts/generate_query_results_in_batch.py \  
              --env dev \
              --list_phrase_file scripts/list_query_file.txt \
              --query_config ocronly
```

Inside the folder scripts/query_results some CSV files should be generated for each of the queries included in list_query_file.txt

**Run the Use Case 4: Query solr using the API**

query endpoint: `curl --location 'http://localhost:8000/query/?query=persons%20and%20corporations&env=dev' --form 'query="'\''\"persons and corporations\"'\''"'`

Output:

`{"id": "chi.67160761", "author": ["Hawaii. Public Utilities Commission"], "title": ["Annual report"]}{"id": "coo.31924007104676", "author": ["Wisconsin. Railroad Commissioners' Department."], "title": ["Biennial report of the Railroad Commissioner of the State of Wisconsin"]}`

status endpoint: `curl --location 'http://localhost:8000/ping/'`
![image](https://github.com/user-attachments/assets/a90ec01a-e7d9-40d4-b4ee-e53d74ed8b20)
